pkg/controller/networkpolicy/validate_test.go

```diff
@@ -1528,14 +1616,236 @@ func TestValidateAntreaGroup(t *testing.T) {
+			validator := NewNetworkPolicyValidator(controller.NetworkPolicyController)
+			actualReason, allowed := validator.validateAntreaGroup(tt.curGroup, tt.oldGroup, tt.operation, authenticationv1.UserInfo{})
+			assert.Equal(t, tt.expectedReason, actualReason)
+			if tt.expectedReason == "" {
+				assert.True(t, allowed)
+			} else {
+				assert.False(t, allowed)
+			}
+		})
+	}
+}
+
+func TestValidateTier(t *testing.T) {
+	tests := []struct {
+		name           string
+		curTier        *crdv1alpha1.Tier
+		oldTier        *crdv1alpha1.Tier
+		operation      admv1.Operation
+		expectedReason string
+	}{
+		{
+			name: "create-tier-pass",
+			curTier: &crdv1alpha1.Tier{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      "tier-priority-3",
+					Namespace: "x",
+				},
+				Spec: crdv1alpha1.TierSpec{
+					Priority: 3,
+				},
+			},
+			operation: admv1.Create,
+		},
+		{
+			name: "create-tier-failed-with-reserved-priority",
+			curTier: &crdv1alpha1.Tier{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      "tier-priority-251",
+					Namespace: "x",
+				},
+				Spec: crdv1alpha1.TierSpec{
+					Priority: 251,
+				},
+			},
+			operation:      admv1.Create,
+			expectedReason: "tier tier-priority-251 priority 251 is reserved",
+		},
+		{
+			name: "delete-tier-pass",
+			oldTier: &crdv1alpha1.Tier{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      "tier-priority-3",
+					Namespace: "x",
+				},
+				Spec: crdv1alpha1.TierSpec{
+					Priority: 3,
+				},
+			},
+			operation: admv1.Delete,
+		},
 	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			_, controller := newController()
+			validator := NewNetworkPolicyValidator(controller.NetworkPolicyController)
+			actualReason, allowed := validator.validateTier(tt.curTier, tt.oldTier, tt.operation, authenticationv1.UserInfo{})
+			assert.Equal(t, tt.expectedReason, actualReason)
+			if tt.expectedReason == "" {
+				assert.True(t, allowed)
+			} else {
+				assert.False(t, allowed)
+			}
+		})
+	}
+}
+
+func TestValidateMulticastIGMP(t *testing.T) {
+	tests := []struct {
+		name           string
+		ingressRules   []crdv1alpha1.Rule
+		egressRules    []crdv1alpha1.Rule
+		expectedReason string
+	}{
+		{
+			name: "igmp-icmp-both-specified",
+			ingressRules: []crdv1alpha1.Rule{
+				{
+					Protocols: []crdv1alpha1.NetworkPolicyProtocol{
+						{
+							ICMP: &crdv1alpha1.ICMPProtocol{
+								ICMPType: &icmpType8,
+								ICMPCode: &icmpCode0,
+							},
+							IGMP: &crdv1alpha1.IGMPProtocol{
+								IGMPType:     &query,
+								GroupAddress: "224.0.0.1",
+							},
+						},
+					},
+					Action: &allowAction,
+				},
+			},
+			egressRules: []crdv1alpha1.Rule{
+				{
+					Protocols: []crdv1alpha1.NetworkPolicyProtocol{
+						{
+							ICMP: &crdv1alpha1.ICMPProtocol{},
+							IGMP: &crdv1alpha1.IGMPProtocol{
+								IGMPType:     &report,
+								GroupAddress: "225.1.2.3",
+							},
+						},
+					},
+					Action: &allowAction,
+				},
+			},
```

Comment:
nit: no need to repeat the same invalid case in both ingress and egress rules. Use ingress rules for one invalid case and use egress rules for another invalid case so it makes sure both ingress and egress validation work.